_episodes/07-b-numpy.md

```diff
@@ -0,0 +1,280 @@
+
+Imagine you have molecular coordinates stored in a numpy array, and you'd like to translate them by a particular distance vector.
+
+~~~
+coordinates = np.array([[0,0,0], [0,1,0],[0,0,1]])
+translation_vector = np.array([2,1,0])
+
+print(coordinates)
+~~~
+{: .language-python}
+
+We would like to move our molecule 2 units in the x direction, 1 unit in the 0 direction, and 0 units in the z direction.
+
+If you were writing a `for` loop to do this, you might write it this way
+
+~~~
+new_coordinates = np.zeros([3,3])
+
+for dim in range(len(translation_vector)):
+    new_coordinates[:, dim] = coordinates[:, dim] + translation_vector[dim]
+
+print(new_coordinates)
+~~~
+
+~~~
+[[2. 1. 0.]
+ [2. 2. 0.]
+ [2. 1. 1.]]
+~~~
+{: .output}
+
+Broadcasting in `numpy` allows us to achieve that with one command, rather than a `for` loop.
+
+~~~
+new_coordinates = coordinates + translation_vector
+
+print(new_coordinates)
+~~~
+{: .language-python}
+
+> ## Exercise
+> Use the features of numpy arrays to subtract the mean of each column from the column data. Leave the `frame` column unchanged. Save your new array as a variable with the name `deviation`
+>> ## Solution
+>> Recall that we already have a variable with the average of each column (`column_averages`). Using broadcasting, we could subtract this vector from our data array. However, since we want to leave frame unchanged, there is one additional step. Since we don't care about the frame average, we will overwrite it with zero.
+>> ~~~
+>> column_averages[0] = 0
+>> deviation = data - column_averages
+>> print(deviation)
+>> ~~~
+>> {: .language-python}
+> {: .solution}
+{: .challenge}
+
+## Optional - Returning to the geometry analysis project
+
+In your geometry analysis project, you had to analyze an xyz file, find the bonds, and print bond lengths.
+
+We can rewrite that project using the features of numpy arrays.  
+
+Recall that a solution given for a function for calculating distances between two points was the following
+
+~~~
+def calculate_distance(rA, rB):
+    """Calculate distance between points A and B"""
+    x_dist = (rA[0] - rB[0]) ** 2
+    y_dist = (rA[1] - rB[1]) ** 2
+    z_dist = (rA[2] - rB[2]) ** 2
+    
+    distance = np.sqrt(x_dist + y_dist + z_dist)
+    
+    return distance
+~~~
+{: .language-python}
+
+This function would work for both lists and numpy arrays, because it does not assume that rA and rB can do something like element-wise subtraction. 
+
+> ## Exercise
+> Using what you've learned about numpy arrays, rewrite the `calculate_distance` function to use the features of numpy arrays.
+>> ## Answer
+>> If rA and rB are both numpy arrays, we can use element-wise subtraction
+>> ~~~
+>>  def calculate_distance(rA, rB):
+>>     """Calculate the distance between points A and B"""
+>>     AB = (rB - rA)**2
```

Comment:
I would add to the docstring to say that both rA and rB have to be numpy arrays.   This would also serve as a good general reminder that if your function requires certain data types that that is a good thing to put in the doc string.